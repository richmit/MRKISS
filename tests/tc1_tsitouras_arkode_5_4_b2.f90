! -*- Mode:F90; Coding:us-ascii-unix; fill-column:129 -*-
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!.H.S.!!
!>
!! @file      tc1_template.f90
!! @author    Mitch Richling http://www.mitchr.me/
!! @brief     Test case.@EOL
!! @std       F2023
!! @see       https://github.com/richmit/MRKISS/
!! @copyright 
!!  @parblock
!!  Copyright (c) 2025, Mitchell Jay Richling <http://www.mitchr.me/> All rights reserved.
!!  
!!  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following
!!  conditions are met:
!!  
!!  1. Redistributions of source code must retain the above copyright notice, this list of conditions, and the following
!!     disclaimer.
!!  
!!  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions, and the following
!!     disclaimer in the documentation and/or other materials provided with the distribution.
!!  
!!  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products
!!     derived from this software without specific prior written permission.
!!  
!!  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
!!  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
!!  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
!!  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
!!  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
!!  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
!!  OF THE POSSIBILITY OF SUCH DAMAGE.
!!  @endparblock
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!.H.E.!!

!----------------------------------------------------------------------------------------------------------------------------------
program tc1_tsitouras_arkode_5_4_b2
  use :: mrkiss_config,                      only: rk, ik
  use :: mrkiss_solvers_wt,                  only: one_step_stab_wt
! use :: mrkiss_eerk_bogacki_shampine_3_2,   only: a, b=>b1, c   ! TCASE_COM: bogacki_shampine_3_2_b1
! use :: mrkiss_eerk_bogacki_shampine_3_2,   only: a, b=>b2, c   ! TCASE_COM: bogacki_shampine_3_2_b2
! use :: mrkiss_eerk_bogacki_shampine_4_5,   only: a, b=>b1, c   ! TCASE_COM: bogacki_shampine_4_5_b1
! use :: mrkiss_eerk_bogacki_shampine_4_5,   only: a, b=>b2, c   ! TCASE_COM: bogacki_shampine_4_5_b2
! use :: mrkiss_eerk_cash_karp_5_4,          only: a, b=>b1, c   ! TCASE_COM: cash_karp_5_4_b1
! use :: mrkiss_eerk_cash_karp_5_4,          only: a, b=>b2, c   ! TCASE_COM: cash_karp_5_4_b2
! use :: mrkiss_eerk_dormand_prince_5_4,     only: a, b=>b1, c   ! TCASE_COM: dormand_prince_5_4_b1
! use :: mrkiss_eerk_dormand_prince_5_4,     only: a, b=>b2, c   ! TCASE_COM: dormand_prince_5_4_b2
! use :: mrkiss_eerk_dormand_prince_7_8,     only: a, b=>b1, c   ! TCASE_COM: dormand_prince_7_8_b1
! use :: mrkiss_eerk_dormand_prince_7_8,     only: a, b=>b2, c   ! TCASE_COM: dormand_prince_7_8_b2
! use :: mrkiss_eerk_fehlberg_4_5,           only: a, b=>b1, c   ! TCASE_COM: fehlberg_4_5_b1
! use :: mrkiss_eerk_fehlberg_4_5,           only: a, b=>b2, c   ! TCASE_COM: fehlberg_4_5_b2
! use :: mrkiss_eerk_fehlberg_7_8,           only: a, b=>b1, c   ! TCASE_COM: fehlberg_7_8_b1
! use :: mrkiss_eerk_fehlberg_7_8,           only: a, b=>b2, c   ! TCASE_COM: fehlberg_7_8_b2
! use :: mrkiss_eerk_heun_euler_2_1,         only: a, b=>b1, c   ! TCASE_COM: heun_euler_2_1_b1
! use :: mrkiss_eerk_heun_euler_2_1,         only: a, b=>b2, c   ! TCASE_COM: heun_euler_2_1_b2
! use :: mrkiss_eerk_sofroniou_spaletta_4_3, only: a, b=>b1, c   ! TCASE_COM: sofroniou_spaletta_4_3_b1
! use :: mrkiss_eerk_sofroniou_spaletta_4_3, only: a, b=>b2, c   ! TCASE_COM: sofroniou_spaletta_4_3_b2
! use :: mrkiss_eerk_tsitouras_arkode_5_4,   only: a, b=>b1, c   ! TCASE_COM: tsitouras_arkode_5_4_b1
  use :: mrkiss_eerk_tsitouras_arkode_5_4,   only: a, b=>b2, c   ! TCASE_COM: tsitouras_arkode_5_4_b2
! use :: mrkiss_eerk_verner_1978_6_5,        only: a, b=>b1, c   ! TCASE_COM: verner_1978_6_5_b1
! use :: mrkiss_eerk_verner_1978_6_5,        only: a, b=>b2, c   ! TCASE_COM: verner_1978_6_5_b2
! use :: mrkiss_eerk_verner_2010_6_5,        only: a, b=>b1, c   ! TCASE_COM: verner_2010_6_5_b1
! use :: mrkiss_eerk_verner_2010_6_5,        only: a, b=>b2, c   ! TCASE_COM: verner_2010_6_5_b2
! use :: mrkiss_eerk_verner_9_8,             only: a, b=>b1, c   ! TCASE_COM: verner_9_8_b1
! use :: mrkiss_eerk_verner_9_8,             only: a, b=>b2, c   ! TCASE_COM: verner_9_8_b2
! use :: mrkiss_erk_euler_1,                 only: a, b, c       ! TCASE_COM: euler_1
! use :: mrkiss_erk_feagin_10,               only: a, b, c       ! TCASE_COM: feagin_10
! use :: mrkiss_erk_knoth_wolke_3,           only: a, b, c       ! TCASE_COM: knoth_wolke_3
! use :: mrkiss_erk_kutta_4,                 only: a, b, c       ! TCASE_COM: kutta_4
! use :: mrkiss_erk_kutta_three_eight_4,     only: a, b, c       ! TCASE_COM: kutta_three_eight_4
! use :: mrkiss_erk_midpoint_2,              only: a, b, c       ! TCASE_COM: midpoint_2
! use :: mrkiss_erk_ralston_2,               only: a, b, c       ! TCASE_COM: ralston_2
! use :: mrkiss_erk_ralston_3,               only: a, b, c       ! TCASE_COM: ralston_3
! use :: mrkiss_erk_ralston_4,               only: a, b, c       ! TCASE_COM: ralston_4

  implicit none
  integer(kind=ik),  parameter :: deq_dim       = 1
  integer(kind=ik),  parameter :: max_steps     = 30
  real(kind=rk),     parameter :: t_end         = 3.0_rk
  real(kind=rk),     parameter :: t_delta       = t_end / (max_steps - 1)
  real(kind=rk),     parameter :: param(1)      = [0.0_rk]
  real(kind=rk),     parameter :: t_iv          = 0.0_rk
  real(kind=rk),     parameter :: y_iv(deq_dim) = [1.0_rk]

  integer(kind=ik) :: step, status
  real(kind=rk)    :: y_delta(deq_dim), y_cv(deq_dim), t_cv, y_tmp(deq_dim)
  integer          :: out_io_stat, out_io_unit

  character(len=*), parameter  :: fmt = "(a40,i5,f23.17,f23.17,f23.17,f23.17)"

  open(newunit=out_io_unit, file="tc1_tsitouras_arkode_5_4_b2.out", form='formatted', action='write', iostat=out_io_stat)
  t_cv = t_iv
  y_cv = y_iv
  do step=1,max_steps
     call ysol(status, y_tmp, t_cv, param)
     write (out_io_unit, fmt=fmt) "tsitouras_arkode_5_4_b2", step, t_cv, y_cv, y_tmp, abs(y_tmp-y_cv)
     call one_step_stab_wt(status, y_delta, eq, t_cv, y_cv, param, a, b, c, t_delta=t_delta)
     t_cv = t_cv + t_delta
     y_cv = y_cv + y_delta
  end do
  close(unit=out_io_unit, status='keep', iostat=out_io_stat)

contains

  subroutine eq(status, dydt, t, y, param)
    integer(kind=ik), intent(out) :: status
    real(kind=rk),    intent(out) :: dydt(:)
    real(kind=rk),    intent(in)  :: t
    real(kind=rk),    intent(in)  :: y(:)
    real(kind=rk),    intent(in)  :: param(:)
    dydt(1) = -2*t*y(1)
    status = 0
  end subroutine eq

  subroutine ysol(status, y, t, param)
    integer(kind=ik), intent(out) :: status
    real(kind=rk),    intent(out) :: y(:)
    real(kind=rk),    intent(in)  :: t
    real(kind=rk),    intent(in)  :: param(:)
    y = exp(-t**2)
    status = 0
  end subroutine ysol

end program tc1_tsitouras_arkode_5_4_b2
