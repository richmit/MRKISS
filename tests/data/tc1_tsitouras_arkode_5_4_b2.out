                 tsitouras_arkode_5_4_b2    1    0.00000000000000000    1.00000000000000000    1.00000000000000000    0.00000000000000000
                 tsitouras_arkode_5_4_b2    2    0.10344827586206896    0.98935550854192689    0.98935551204517247    0.00000000350324558
                 tsitouras_arkode_5_4_b2    3    0.20689655172413793    0.95809708170912777    0.95809706746156065    0.00000001424756713
                 tsitouras_arkode_5_4_b2    4    0.31034482758620691    0.90817893586171394    0.90817888305953165    0.00000005280218229
                 tsitouras_arkode_5_4_b2    5    0.41379310344827586    0.84263229382813709    0.84263218538655249    0.00000010844158460
                 tsitouras_arkode_5_4_b2    6    0.51724137931034486    0.76526093091593361    0.76526075695437135    0.00000017396156227
                 tsitouras_arkode_5_4_b2    7    0.62068965517241381    0.68027694165574781    0.68027670256829587    0.00000023908745195
                 tsitouras_arkode_5_4_b2    8    0.72413793103448276    0.59192505381021898    0.59192476213508860    0.00000029167513038
                 tsitouras_arkode_5_4_b2    9    0.82758620689655171    0.50414150024498028    0.50414118062702185    0.00000031961795843
                 tsitouras_arkode_5_4_b2   10    0.93103448275862066    0.42028405996337859    0.42028374684482200    0.00000031311855658
                 tsitouras_arkode_5_4_b2   11    1.03448275862068972    0.34295576678630135    0.34295549998583796    0.00000026680046339
                 tsitouras_arkode_5_4_b2   12    1.13793103448275867    0.27392898689702760    0.27392880579018886    0.00000018110683875
                 tsitouras_arkode_5_4_b2   13    1.24137931034482762    0.21416205162874374    0.21416198904547337    0.00000006258327037
                 tsitouras_arkode_5_4_b2   14    1.34482758620689657    0.16388968877285620    0.16388976584779275    0.00000007707493654
                 tsitouras_arkode_5_4_b2   15    1.44827586206896552    0.12276236438162094    0.12276258738810412    0.00000022300648318
                 tsitouras_arkode_5_4_b2   16    1.55172413793103448    0.09000844345393562    0.09000880318777127    0.00000035973383565
                 tsitouras_arkode_5_4_b2   17    1.65517241379310343    0.06459598301561879    0.06459645671042596    0.00000047369480717
                 tsitouras_arkode_5_4_b2   18    1.75862068965517238    0.04537658908663728    0.04537714430590826    0.00000055521927098
                 tsitouras_arkode_5_4_b2   19    1.86206896551724133    0.03120053459906319    0.03120113419753158    0.00000059959846839
                 tsitouras_arkode_5_4_b2   20    1.96551724137931028    0.02099886397736509    0.02099947112374686    0.00000060714638176
                 tsitouras_arkode_5_4_b2   21    2.06896551724137945    0.01383352326133697    0.01383410565439463    0.00000058239305766
                 tsitouras_arkode_5_4_b2   22    2.17241379310344840    0.00892015989507559    0.00892069261069119    0.00000053271561560
                 tsitouras_arkode_5_4_b2   23    2.27586206896551735    0.00563008319041591    0.00563054996455342    0.00000046677413751
                 tsitouras_arkode_5_4_b2   24    2.37931034482758630    0.00347823352196940    0.00347862660438811    0.00000039308241871
                 tsitouras_arkode_5_4_b2   25    2.48275862068965525    0.00210331208397421    0.00210363102336487    0.00000031893939066
                 tsitouras_arkode_5_4_b2   26    2.58620689655172420    0.00124494147504400    0.00124519129312524    0.00000024981808125
                 tsitouras_arkode_5_4_b2   27    2.68965517241379315    0.00072126263632801    0.00072145182940382    0.00000018919307580
                 tsitouras_arkode_5_4_b2   28    2.79310344827586210    0.00040901204767151    0.00040915075620006    0.00000013870852855
                 tsitouras_arkode_5_4_b2   29    2.89655172413793105    0.00022702602339296    0.00022712457769360    0.00000009855430064
                 tsitouras_arkode_5_4_b2   30    3.00000000000000000    0.00012334188234562    0.00012340980408668    0.00000006792174106
